template.go

```diff
@@ -121,16 +130,31 @@ func NewTemplate(i TemplateInput) *Template {
 	t.dirty = make(drainableChan, 1)
 	t.Notify(nil) // prime template as needing to be run
 
+	t.name = i.Name
+	if t.name == "" {
+		t.name = randStr(randNameLength)
```

Comment:
The default value for the template name may be difficult to back track which template it is associated with in relation to the template ID. And since the name is autogenerated, it would not be clear which identifier to use for logging purposes: template name or ID. They kinda end up as two miscellaneous values.

What are your thoughts around having the default name fallback to the template ID? 

In this case, the hash of the content and would only allow 1 nameless template of the same content and the second nameless result in the same template from hcat's perspective. It would preserve an option for the same behavior as now, and offer the new feature to have duplicate templates using different names. The developer could choose which design is appropriate for their application

Reply:
I went back and forth on this myself. I thought the API would be simpler if it used the name-suffix by default to keep the templates/notifiers distinct between instances. As opposed to having it be optional and needing to document why you might want to set it. OTOH, the code would obviously be simpler the other way. 

I assumed CTS would use explicit names, so it wouldn't be an issue. Do you see it being one?

... https://www.youtube.com/watch?v=7QEB38D0pu4 ...

Pondering this more I guess I'm leaning toward your line of thought. Given CTS doesn't need (or want) the random names and all the other current use cases are fine with just the hex-id as the id... I think I'm trying to predict what will be useful the future, which is pretty much always a bad call. 

TLDR; I'm going to strip out the random default, add some documentation and see how things go.

Reply:
LOL the video caught me off guard! 😂 this was a fun comment to return back to.

Yeah, CTS would use explicit names and the ID+name suffix wouldn't be a problem.

For the Consul Template logging/(future)metric use-case, I anticipate 2 things following the proposal:
1. Logging/metrics based on the name instead of ID would be ideal to keep the same template (updated after reloading changes from disk) would be attributed to the same aggregated bin.
2. Different and unrelated templates can have the same name and it would be up to the caller to unsure it's unique.